src/main/kotlin/com/exactpro/th2/check1/metrics/BufferMetric.kt

```diff
@@ -0,0 +1,123 @@
+            .build("th2_check1_actual_cache_size", "The actual size of the cache")
+            .labelNames(DEFAULT_SESSION_ALIAS_LABEL_NAME, DEFAULT_DIRECTION_LABEL_NAME)
+            .register()
+    private val requiredCacheMemorySizeMetric: Gauge = Gauge
+            .build("th2_check1_required_cache_memory_size", "Required cache memory size without active cache size")
+            .register()
+
+    private lateinit var lastReceivedMessageMetricTask: Disposable
+    private val lastReceivedMessages: Queue<Long> = ConcurrentLinkedQueue<Long>()
+    private val bufferMessagesSizeBySessionKey: MutableMap<SessionKey, LinkedList<Long>> = ConcurrentHashMap()
+    private var maxBufferSize: Int = -1
+
+    fun configure(configuration: Check1Configuration) {
+        this.maxBufferSize = configuration.messageCacheSize
+        this.lastReceivedMessageMetricTask = createObservableInterval(configuration.messageSizeCollectionTimeout)
+    }
+
+    fun processMessage(sessionKey: SessionKey, message: Message) {
+        val calculatedMessageSize = calculateMessageSize(message)
+        val labels = arrayOf(sessionKey.sessionAlias, sessionKey.direction.name)
+
+        bufferMessagesSizeBySessionKey.compute(sessionKey) { _, bufferedMessageSize ->
+            if (bufferedMessageSize == null) {
+                incrementStats(calculatedMessageSize, labels)
+                return@compute LinkedList<Long>().apply { add(calculatedMessageSize) }
+            }
+
+            if (bufferedMessageSize.size >= maxBufferSize) {
+                decrementStats(bufferedMessageSize.poll(), labels)
+            }
+
+            bufferedMessageSize.add(calculatedMessageSize)
+            incrementStats(calculatedMessageSize, labels)
+
+            return@compute bufferedMessageSize
+        }
+
+        lastReceivedMessages.add(calculatedMessageSize)
+        requiredCacheMemorySizeMetric.set(calculateRequiredMemory())
+    }
+
+    override fun close() {
+        lastReceivedMessageMetricTask.dispose()
+    }
+
+    private fun calculateMessageSize(message: Message): Long {
+        val parsedInstance = GraphLayout.parseInstance(message)
+        if (logger.isTraceEnabled) {
+            logger.trace("Foot print for message with id: '{}'\n{}", message.metadata.id.toJson(), parsedInstance.toFootprint())
+        }
+        return parsedInstance.totalSize()
+    }
+
+    private fun createObservableInterval(messageSizeCollectionTimeout: Long) =
+            Observable.interval(messageSizeCollectionTimeout, TimeUnit.MINUTES)
+                    .subscribe {
+                        averageReceivedMessagesSizeMetric.set(lastReceivedMessages.average())
+                        lastReceivedMessages.clear()
+                    }
+
+    private fun calculateRequiredMemory(): Double {
+        return bufferMessagesSizeBySessionKey.values.sumByDouble {
+            val averageMessageSize = it.average()
+            val leftBufferCapacity = maxBufferSize - it.size
```

Comment:
What do we compute here? How much memory will be additionally consumed when we fill up the buffer?

Maybe we should calculate how much memory needs the buffer when it will be filled up (averageMessageSize * maxBufferSize)? 

Also, I think we can compute that using Grafana's functions